### PR TITLE
RA-1269 environments

### PR DESCRIPTION
**Pull request by JoakimHagen**

### CODE REVIEW

File: api/applications/applications_controller_test.go [hunk 1]
```diff
@@ -418,6 +418,21 @@ func TestGetApplication_WithEnvironments(t *testing.T) {
 		WithEnvironment("prod", "release").
 		WithEnvironment(anyOrphanedEnvironment, "feature"))
 
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/applications/applications_controller_test.go]: I think you should not need to do this. I think you have misunderstood the builder. I think the application builder should take care of construction of an environment. A tester should not need to create and apply these environments, unless it is needed for testing. Also the opposite should take place when we want to test an RE. That is a environmentBuilder should construct a registration if not provided

File: .vscode/launch.json [hunk 2]
```diff
@@ -18,6 +18,15 @@
             "args": [
                 "--useOutClusterClient=false"
             ]
+        },
```
Comment (ingeknudsen) [inline .vscode/launch.json]: Why do you keep adding these lauch configs. Doesnt go tools work with you vscode? You should see a link that will allow you to run a single test or all tests in a package

File: api/applications/applications_controller_test.go [hunk 3]
```diff
@@ -418,6 +418,21 @@ func TestGetApplication_WithEnvironments(t *testing.T) {
 		WithEnvironment("prod", "release").
```
Comment (ingeknudsen) [inline api/applications/applications_controller_test.go]: See below comment. So I think I would have removed WithEnvironment(anyOrphanedEnvironment, "feature"))

File: api/applications/applications_controller_test.go [hunk 4]
```diff
@@ -418,6 +418,21 @@ func TestGetApplication_WithEnvironments(t *testing.T) {
...
+		NewEnvironmentBuilder().
+		WithAppName(anyAppName).WithAppLabel().
+		WithEnvironmentName("dev"))
+
+	commonTestUtils.ApplyEnvironment(builders.
+		NewEnvironmentBuilder().
+		WithAppName(anyAppName).WithAppLabel().
+		WithEnvironmentName("prod"))
+
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/applications/applications_controller_test.go]: This is the only ApplyEnvironment I would have left here

File: api/applications/applications_controller_test.go [hunk 5]
```diff
@@ -642,6 +657,16 @@ func TestHandleTriggerPipeline_ForNonMappedAndMappedAndMagicBranchEnvironment_Jo
 		WithEnvironment("dev", "dev").
 		WithEnvironment("prod", "release"))
 
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/applications/applications_controller_test.go]: These two ApplyEnvironment I would have removed entirely

File: api/applications/applications_controller_test.go [hunk 6]
```diff
@@ -856,6 +881,11 @@ func TestGetApplication_WithAppAlias_ContainsAppAlias(t *testing.T) {
 				builders.NewDeployComponentBuilder().
 					WithName("backend")))
 
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/applications/applications_controller_test.go]: This I would also have removed, as it adds no value

File: api/environments/environment_controller_test.go [hunk 7]
```diff
@@ -274,6 +274,16 @@ func TestGetEnvironmentSummary_OrphanedEnvironmentWithDash_OrphanedEnvironmentIs
 		WithEnvironment("dev", "master").
 		WithEnvironment(anyOrphanedEnvironment, "feature"))
 
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/environments/environment_controller_test.go]: Would have deleted the line:
WithEnvironment(anyOrphanedEnvironment, "feature")

... and this ApplyEnviornment

File: api/environments/environment_controller_test.go [hunk 8]
```diff
@@ -274,6 +274,16 @@ func TestGetEnvironmentSummary_OrphanedEnvironmentWithDash_OrphanedEnvironmentIs
...
+		NewEnvironmentBuilder().
+		WithAppName(anyAppName).WithAppLabel().
+		WithEnvironmentName("dev"))
+
+	commonTestUtils.ApplyEnvironment(builders.
```
Comment (ingeknudsen) [inline api/environments/environment_controller_test.go]: This is ok to keep

Comment (ingeknudsen) [reply api/environments/environment_controller_test.go]: And then I would have deleted:
commonTestUtils.ApplyApplicationUpdate

File: api/environments/environment_handler.go [hunk 9]
```diff
@@ -170,68 +156,114 @@ func (eh EnvironmentHandler) getConfigurationStatusOfNamespace(namespace string)
...
+	_, err := eh.client.CoreV1().Namespaces().Get(uniqueName, metav1.GetOptions{})
+	return err == nil, err
+}
+
 func getNamespacesInConfig(radixApplication *v1.RadixApplication) map[string]bool {
 	namespacesInConfig := make(map[string]bool)
 	for _, environment := range radixApplication.Spec.Environments {
-		environmentNamespace := k8sObjectUtils.GetEnvironmentNamespace(radixApplication.Name, environment.Name)
-		namespacesInConfig[environmentNamespace] = true
+		uniqueName := k8sObjectUtils.GetEnvironmentNamespace(radixApplication.Name, environment.Name)
```
Comment (ingeknudsen) [inline api/environments/environment_handler.go]: I guess all references to namespace here is wrong. I.e. getNamespacesInConfig, namespacesInConfig etc.

File: api/environments/environment_handler.go [hunk 10]
```diff
@@ -170,68 +156,114 @@ func (eh EnvironmentHandler) getConfigurationStatusOfNamespace(namespace string)
...
+	for _, re := range radixEnvironments.Items {
+		if isOrphaned(re.Name, namespacesInConfig) {
+			envNames = append(envNames, re.Spec.EnvName)
+		}
+	}
+
+	return envNames
+}
+
+func (eh EnvironmentHandler) namespaceExists(uniqueName string) (bool, error) {
```
Comment (ingeknudsen) [inline api/environments/environment_handler.go]: Should probably be named environmentExists

File: .vscode/launch.json (hunk 2 above)
Comment (JoakimHagen) [reply .vscode/launch.json:21]: It's the easiest way for me to test. The terminal takes time writing in and copy pasting the full project/package path

File: api/applications/applications_controller_test.go (hunk 4 above)
Comment (JoakimHagen) [reply api/applications/applications_controller_test.go]: I removed them all. I'm not sure what you want to replace with the separate call to ApplyEnvironment now that Application can create it.

File: .vscode/launch.json (hunk 2 above)
Comment (JoakimHagen) [reply .vscode/launch.json]: Removed them in this PR

**Pull request merged**